src/commands/interface-manager/MatrixPromptUX.ts

```diff
@@ -0,0 +1,171 @@
+
+// Returns true if the listener should be kept.
+type ReactionPromptListener = (presentation: any) => boolean|void;
+
+// Instead of providing a map of reaciton keys to presentations, should instead
+// there be provided an object that can quickly be interned and uninterened from the table
+// by testing EQ? It's important to wrap the map if you consider that
+// multiple invocations can have seperate listeners but share the same map.
+class ReactionPromptRecord {
+    constructor(
+        public readonly presentationByReaction: PresentationByReactionKey,
+        public readonly listener: ReactionPromptListener,
+    ) {
+        // nothing to do.
+    }
+}
+
+class ReactionHandler {
+    private readonly promptRecordByEvent: Map<string/*event id*/, Set<ReactionPromptRecord>> = new Map();
+
+    constructor(
+        matrixEmitter: MatrixEmitter
+    ) {
+        matrixEmitter.on('room.event', this.handleEvent.bind(this))
+    }
+
+    private addPresentationsForEvent(eventId: string, promptRecord: ReactionPromptRecord): void {
+        const promptRecords = (() => {
+            let entry = this.promptRecordByEvent.get(eventId);
+            if (entry === undefined) {
+                entry = new Set();
+                this.promptRecordByEvent.set(eventId, entry);
+            }
+            return entry;
+        })();
+        promptRecords.add(promptRecord);
+    }
+
+    private removePromptRecordForEvent(eventId: string, promptRecord: ReactionPromptRecord): void {
+        const promptRecords = this.promptRecordByEvent.get(eventId);
+        if (promptRecords !== undefined) {
+            promptRecords.delete(promptRecord);
+            if (promptRecords.size === 0) {
+                this.promptRecordByEvent.delete(eventId);
+            }
+        }
+    }
+
+    private handleEvent(roomId: string, event: { event_id: string, type: string, content: any }): void {
+        // Horrid, would be nice to have some pattern matchy thingo
+        if (event.type !== 'm.reaction') {
+            return;
+        }
+        const relatesTo = event['content']?.['m.relates_to'];
+        if (relatesTo === undefined) {
+            return;
+        }
+        if (relatesTo['rel_type'] !== 'm.annotation') {
+            return;
+        }
+        const relatedEventId = relatesTo['event_id'];
+        const reactionKey = relatesTo['key'];
+        if (!(typeof relatedEventId === 'string' && typeof reactionKey === 'string')) {
+            return;
+        }
+        const entry = this.promptRecordByEvent.get(relatedEventId);
+        if (entry !== undefined) {
+            for (const record of entry) {
+                const presentation = record.presentationByReaction.get(reactionKey);
+                if (presentation === undefined) {
+                    // FIXME: Should this be WARN? Technically the prompt should fail as saying
+                    // that the reaction wasn't understood.
```

Comment:
Hmmmm I feel like i just deleted this comment thinking it was asking for a warn rather than a fail

Reply:
Honestly in the grand scheme of things this does not matter